### PR TITLE
MAINT: Update for latest VTK

### LINKED ISSUE #1049: Build hangs on Fedora 34
When trying to build the mayavi package on Fedora 34, the build hangs:
```
+ SPHINXBUILD=/usr/bin/sphinx-build-3
+ /usr/bin/python3 setup.py build '--executable=/usr/bin/python3 -s'
running build
----------------------------------------------------------------------
Building TVTK classes...
```
strace shows:
```
write(1, "Building TVTK classes... ", 25Building TVTK classes... ) = 25
getcwd("/home/orion/fedora/Mayavi/mayavi-4.7.3", 1024) = 39
chdir("/home/orion/fedora/Mayavi/mayavi-4.7.3/tvtk") = 0
getcwd("/home/orion/fedora/Mayavi/mayavi-4.7.3/tvtk", 1024) = 44
getpid()                                = 4162
getrandom("\xfc\xeb\x3e\x31\x0f\x5a\x98\xe3\x06\xea\x9a\xbd\x4e\xbc\x86\x26\xb8\x75\x76\x37\xbd\xe5\xc5\xec\xb4\xe2\xf3\xe7\x6c\xbf\xe5\x23"..., 2496, GRND_NONBLOCK) = 2496
openat(AT_FDCWD, "/tmp/cpp6pyan", O_RDWR|O_CREAT|O_EXCL|O_NOFOLLOW|O_CLOEXEC, 0600) = 3
newfstatat(3, "", {st_mode=S_IFREG|0600, st_size=0, ...}, AT_EMPTY_PATH) = 0
ioctl(3, TCGETS, 0x7fff5f6f1240)        = -1 ENOTTY (Inappropriate ioctl for device)
lseek(3, 0, SEEK_CUR)                   = 0
write(3, "blat", 4)                     = 4
close(3)                                = 0
unlink("/tmp/cpp6pyan")                 = 0
getpid()                                = 4162
getrandom("\xd3\x99\x13\xec\x46\xbf\x6c\x35\x6b\x9f\xdf\x64\xfc\x27\xca\x02\xdd\xd4\x6a\x72\x3d\x07\xd5\x2f\x28\x9c\x24\xa6\x98\x43\x3d\x3b"..., 2496, GRND_NONBLOCK) = 2496
mkdir("/tmp/tmpqcxiz7h2", 0700)         = 0
newfstatat(AT_FDCWD, "/tmp/tmpqcxiz7h2/tvtk_classes", 0x7fff5f6f1760, 0) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/tmp/tmpqcxiz7h2", {st_mode=S_IFDIR|0700, st_size=40, ...}, 0) = 0
mkdir("/tmp/tmpqcxiz7h2/tvtk_classes", 0777) = 0
mmap(NULL, 262144, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f9a6e524000
mmap(NULL, 262144, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f9a6e4e4000
mmap(NULL, 262144, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f9a6e4a4000
mmap(NULL, 262144, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f9a6e464000
mmap(NULL, 262144, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f9a6e424000
mmap(NULL, 262144, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f9a6e3e4000
openat(AT_FDCWD, "/tmp/tmpqcxiz7h2/tvtk_classes/__init__.py", O_WRONLY|O_CREAT|O_TRUNC|O_CLOEXEC, 0666) = 3
newfstatat(3, "", {st_mode=S_IFREG|0644, st_size=0, ...}, AT_EMPTY_PATH) = 0
ioctl(3, TCGETS, 0x7fff5f6f1a60)        = -1 ENOTTY (Inappropriate ioctl for device)
lseek(3, 0, SEEK_CUR)                   = 0
ioctl(3, TCGETS, 0x7fff5f6f1860)        = -1 ENOTTY (Inappropriate ioctl for device)
lseek(3, 0, SEEK_CUR)                   = 0
close(3)                                = 0
openat(AT_FDCWD, "/tmp/tmpqcxiz7h2/tvtk_classes/vtk_version.py", O_WRONLY|O_CREAT|O_TRUNC|O_CLOEXEC, 0666) = 3
newfstatat(3, "", {st_mode=S_IFREG|0644, st_size=0, ...}, AT_EMPTY_PATH) = 0
ioctl(3, TCGETS, 0x7fff5f6f1a60)        = -1 ENOTTY (Inappropriate ioctl for device)
lseek(3, 0, SEEK_CUR)                   = 0
ioctl(3, TCGETS, 0x7fff5f6f1860)        = -1 ENOTTY (Inappropriate ioctl for device)
lseek(3, 0, SEEK_CUR)                   = 0
write(3, "vtk_build_version = '9.0'\nvtk_bu"..., 70) = 70
close(3)                                = 0
openat(AT_FDCWD, "/tmp/tmpqcxiz7h2/tvtk_classes/tvtk_helper.py", O_WRONLY|O_CREAT|O_TRUNC|O_CLOEXEC, 0666) = 3
newfstatat(3, "", {st_mode=S_IFREG|0644, st_size=0, ...}, AT_EMPTY_PATH) = 0
ioctl(3, TCGETS, 0x7fff5f6f1a60)        = -1 ENOTTY (Inappropriate ioctl for device)
lseek(3, 0, SEEK_CUR)                   = 0
ioctl(3, TCGETS, 0x7fff5f6f1860)        = -1 ENOTTY (Inappropriate ioctl for device)
lseek(3, 0, SEEK_CUR)                   = 0
```
This never seems to complete and FD 3 remains open.  It continues with FD4:
```
openat(AT_FDCWD, "/tmp/tmpqcxiz7h2/tvtk_classes/object_base.py", O_WRONLY|O_CREAT|O_TRUNC|O_CLOEXEC, 0666) = 4
newfstatat(4, "", {st_mode=S_IFREG|0644, st_size=0, ...}, AT_EMPTY_PATH) = 0
ioctl(4, TCGETS, 0x7fff5f6f18f0)        = -1 ENOTTY (Inappropriate ioctl for device)
lseek(4, 0, SEEK_CUR)                   = 0
lseek(4, 0, SEEK_CUR)                   = 0
write(4, "# Automatically generated code: "..., 7967) = 7967
write(4, "            traits_view = \\\n    "..., 432) = 432
close(4)                                = 0
```
until:
```
openat(AT_FDCWD, "/tmp/tmpqcxiz7h2/tvtk_classes/web_application.py", O_WRONLY|O_CREAT|O_TRUNC|O_CLOEXEC, 0666) = 4
newfstatat(4, "", {st_mode=S_IFREG|0644, st_size=0, ...}, AT_EMPTY_PATH) = 0
ioctl(4, TCGETS, 0x7fff5f6f18f0)        = -1 ENOTTY (Inappropriate ioctl for device)
lseek(4, 0, SEEK_CUR)                   = 0
lseek(4, 0, SEEK_CUR)                   = 0
rt_sigprocmask(SIG_BLOCK, ~[], [], 8)   = 0
clone(child_stack=0x7f9a6e3e2ef0, flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTID, parent_tid=[4171], tls=0x7f9a6e3e3640, child_tidptr=0x7f9a6e3e3910) = 4171
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
rt_sigprocmask(SIG_BLOCK, ~[], [], 8)   = 0
clone(child_stack=0x7f9a6dbe1ef0, flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTID, parent_tid=[4172], tls=0x7f9a6dbe2640, child_tidptr=0x7f9a6dbe2910) = 4172
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
rt_sigprocmask(SIG_BLOCK, ~[], [], 8)   = 0
clone(child_stack=0x7f9a6d3e0ef0, flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTID, parent_tid=[4173], tls=0x7f9a6d3e1640, child_tidptr=0x7f9a6d3e1910) = 4173
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
futex(0x560329578cc0, FUTEX_WAIT_BITSET_PRIVATE|FUTEX_CLOCK_REALTIME, 0, NULL, FUTEX_BITSET_MATCH_ANYstrace: Process 4172 attached
strace: Process 4173 attached
strace: Process 4171 attached
 <unfinished ...>
[pid  4172] set_robust_list(0x7f9a6dbe2920, 24) = 0
[pid  4172] rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
[pid  4172] futex(0x560329578cc0, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid  4162] <... futex resumed>)        = 0
[pid  4172] <... futex resumed>)        = 1
[pid  4162] futex(0x560329578c70, FUTEX_WAKE_PRIVATE, 1) = 0
[pid  4162] futex(0x7f9a6d3e1910, FUTEX_WAIT_BITSET|FUTEX_CLOCK_REALTIME, 4173, NULL, FUTEX_BITSET_MATCH_ANY <unfinished ...>
[pid  4172] madvise(0x7f9a6d3e2000, 8368128, MADV_DONTNEED) = 0
[pid  4172] exit(0)                     = ?
[pid  4172] +++ exited with 0 +++
[pid  4173] set_robust_list(0x7f9a6d3e1920, 24 <unfinished ...>
[pid  4171] set_robust_list(0x7f9a6e3e3920, 24 <unfinished ...>
[pid  4173] <... set_robust_list resumed>) = 0
[pid  4171] <... set_robust_list resumed>) = 0
[pid  4173] rt_sigprocmask(SIG_SETMASK, [],  <unfinished ...>
[pid  4171] rt_sigprocmask(SIG_SETMASK, [],  <unfinished ...>
[pid  4173] <... rt_sigprocmask resumed>NULL, 8) = 0
[pid  4171] <... rt_sigprocmask resumed>NULL, 8) = 0
[pid  4173] futex(0x560329578d58, FUTEX_WAIT_BITSET_PRIVATE|FUTEX_CLOCK_REALTIME, 0, NULL, FUTEX_BITSET_MATCH_ANY <unfinished ...>
[pid  4171] futex(0x560329578d58, FUTEX_WAIT_BITSET_PRIVATE|FUTEX_CLOCK_REALTIME, 0, NULL, FUTEX_BITSET_MATCH_ANY
```
Then we seem to hit a deadlock.  This is using python 3.9.5.

### PR DESCRIPTION
**Pull request by larsoner**
Closes #1049

@opoplawski can you see if this fixes your build hang on Fedora 34? It at least "fixes" the hang for me on Windows. Plotting still doesn't work there (`mlab.test_plot3d()`) despite pyvistaqt.BackgroundPlotter working okay, so there is more to fix. But this might already be good enough progress in that it will prevent total `pip install` hangs for people using the VTK pre wheels. (They'll have a problem later when they try to actually plot, but that's at least easier to diagnose than an [empty build log that took an hour of CI time](https://dev.azure.com/mne-tools/mne-python/_build/results?buildId=13871&view=logs&j=a017e066-62ca-5289-ad0b-8f57c84a089f&t=13529fc1-567a-5a12-e0e9-1de701345c43)).

### CODE REVIEW

**Comment by larsoner** [comment]
It would also be good to add a CI job that tests with `pip install --upgrade --pre vtk`, but that should be another PR since I'm not really familiar with the CI setup here

**Comment by opoplawski** [comment]
@larsoner It seems to fix the hang I was seeing, thanks.

**Comment by mjg** [comment]
On Fedora 34 with Mayavi-4.7.3 including this fix here, vtk-9.0.1 and python 3.9.5, Mayavi2 (the app) now starts and plots but throws some error (see below). `from mayavi import mlab` still core dumps (even with `ETS_TOOLKIT=wx`).
```
2021-06-29 12:39:44.250 (  40,375s) [        FD47C740]     vtkOpenGLState.cxx:505   WARN| Error glBindFramebuffer1 OpenGL errors detected
  0 : (1282) Invalid operation

 with stack trace of
0x7f8aee7ae67e : ??? [(???) ???:-1]
0x7f8aee7af3a2 : vtksys::SystemInformation::GetProgramStack[abi:cxx11](int, int) [(libvtksys.so.1) ???:-1]
0x7f8aec605089 : ??? [(???) ???:-1]
0x7f8aec606b0e : vtkOpenGLState::vtkBindFramebuffer(unsigned int, vtkOpenGLFramebufferObject*) [(libvtkRenderingOpenGL2.so.1) ???:-1]
0x7f8aee2e97fd : vtkRenderWindow::Render() [(libvtkRenderingCore.so.1) ???:-1]
0x7f8aec5e7582 : vtkOpenGLRenderWindow::Render() [(libvtkRenderingOpenGL2.so.1) ???:-1]
0x7f8aec6a6e47 : vtkXOpenGLRenderWindow::Render() [(libvtkRenderingOpenGL2.so.1) ???:-1]
0x7f8aea7667ad : ??? [(???) ???:-1]
0x7f8afd8e0f68 : ??? [(???) ???:-1]
0x7f8afd8d2de7 : _PyObject_MakeTpCall [(libpython3.9.so.1.0) ???:-1]
0x7f8afd8cfcee : _PyEval_EvalFrameDefault [(libpython3.9.so.1.0) ???:-1]
0x7f8afd8d7753 : ??? [(???) ???:-1]
0x7f8afd8cf73f : _PyEval_EvalFrameDefault [(libpython3.9.so.1.0) ???:-1]
0x7f8afd8d7753 : ??? [(???) ???:-1]
0x7f8afd8dfa92 : ??? [(???) ???:-1]
0x7f8ab8b742fa : ??? [(???) ???:-1]
0x7f8ab8b743e3 : ??? [(???) ???:-1]
0x7f8ab3d6f649 : ??? [(???) ???:-1]
0x7f8ae865885e : QWidget::event(QEvent*) [(libQt5Widgets.so.5) ???:-1]
0x7f8ab3d7095b : ??? [(???) ???:-1]
0x7f8ae8617e73 : QApplicationPrivate::notify_helper(QObject*, QEvent*) [(libQt5Widgets.so.5) ???:-1]
0x7f8ab3b385b6 : ??? [(???) ???:-1]
0x7f8ae7b89f48 : QCoreApplication::notifyInternal2(QObject*, QEvent*) [(libQt5Core.so.5) ???:-1]
0x7f8ae865093a : QWidgetPrivate::sendPaintEvent(QRegion const&) [(libQt5Widgets.so.5) ???:-1]
0x7f8ae86514df : QWidgetPrivate::drawWidget(QPaintDevice*, QRegion const&, QPoint const&, QFlags<QWidgetPrivate::DrawWidgetFlag>, QPainter*, QWidgetRepaintManager*) [(libQt5W
idgets.so.5) ???:-1]
0x7f8ae8651d52 : QWidgetPrivate::paintOnScreen(QRegion const&) [(libQt5Widgets.so.5) ???:-1]
0x7f8ae8651e88 : QWidgetPrivate::syncBackingStore() [(libQt5Widgets.so.5) ???:-1]
0x7f8ae8658f8d : QWidget::event(QEvent*) [(libQt5Widgets.so.5) ???:-1]
0x7f8ab3d7095b : ??? [(???) ???:-1]
0x7f8ae8617e73 : QApplicationPrivate::notify_helper(QObject*, QEvent*) [(libQt5Widgets.so.5) ???:-1]
0x7f8ab3b385b6 : ??? [(???) ???:-1]
0x7f8ae7b89f48 : QCoreApplication::notifyInternal2(QObject*, QEvent*) [(libQt5Core.so.5) ???:-1]
0x7f8ae7b8cc76 : QCoreApplicationPrivate::sendPostedEvents(QObject*, int, QThreadData*) [(libQt5Core.so.5) ???:-1]
0x7f8ae7bd6c57 : ??? [(???) ???:-1]
0x7f8aec0374cf : g_main_context_dispatch [(libglib-2.0.so.0) ???:-1]
0x7f8aec08b4e8 : ??? [(???) ???:-1]
0x7f8aec034c03 : g_main_context_iteration [(libglib-2.0.so.0) ???:-1]
0x7f8ae7bd66f8 : QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) [(libQt5Core.so.5) ???:-1]
0x7f8ae7b889b2 : QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) [(libQt5Core.so.5) ???:-1]
0x7f8ae7b90544 : QCoreApplication::exec() [(libQt5Core.so.5) ???:-1]
0x7f8ab3b39b2e : ??? [(???) ???:-1]
0x7f8afd8e0f68 : ??? [(???) ???:-1]
0x7f8afd8d2de7 : _PyObject_MakeTpCall [(libpython3.9.so.1.0) ???:-1]
0x7f8afd8cfcee : _PyEval_EvalFrameDefault [(libpython3.9.so.1.0) ???:-1]
0x7f8afd8d7753 : ??? [(???) ???:-1]
0x7f8afd8caa5a : _PyEval_EvalFrameDefault [(libpython3.9.so.1.0) ???:-1]
0x7f8afd8d7753 : ??? [(???) ???:-1]
0x7f8afd8df9f1 : ??? [(???) ???:-1]
0x7f8afd8cf73f : _PyEval_EvalFrameDefault [(libpython3.9.so.1.0) ???:-1]
0x7f8afd8d7753 : ??? [(???) ???:-1]
0x7f8afd8df9f1 : ??? [(???) ???:-1]
0x7f8afd8cf73f : _PyEval_EvalFrameDefault [(libpython3.9.so.1.0) ???:-1]
0x7f8afd8c96dd : ??? [(???) ???:-1]
0x7f8afd8d745e : _PyFunction_Vectorcall [(libpython3.9.so.1.0) ???:-1]
0x7f8afd8df9f1 : ??? [(???) ???:-1]
0x7f8afd8cf73f : _PyEval_EvalFrameDefault [(libpython3.9.so.1.0) ???:-1]
0x7f8afd8d7753 : ??? [(???) ???:-1]
0x7f8afd8caa5a : _PyEval_EvalFrameDefault [(libpython3.9.so.1.0) ???:-1]
0x7f8afd8c96dd : ??? [(???) ???:-1]
0x7f8afd9460e5 : _PyEval_EvalCodeWithName [(libpython3.9.so.1.0) ???:-1]
0x7f8afd94607d : PyEval_EvalCodeEx [(libpython3.9.so.1.0) ???:-1]
0x7f8afd94602f : PyEval_EvalCode [(libpython3.9.so.1.0) ???:-1]
0x7f8afd9722e4 : ??? [(???) ???:-1]
0x7f8afd96e566 : ??? [(???) ???:-1]
0x7f8afd84621a : ??? [(???) ???:-1]
0x7f8afd968ac3 : PyRun_SimpleFileExFlags [(libpython3.9.so.1.0) ???:-1]
0x7f8afd965ec8 : Py_RunMain [(libpython3.9.so.1.0) ???:-1]
0x7f8afd938add : Py_BytesMain [(libpython3.9.so.1.0) ???:-1]
0x7f8afd619b75 : __libc_start_main [(libc.so.6) ???:-1]
0x55f2fbd2509e : _start [(python3) ???:-1]

```

**Comment by larsoner** [comment]
VTK just released 9.0.2, and all my `pip install` commands are stalled on Windows. I'm going to work around it by using my branch, but I expect this to be a (somewhat hard to diagnose) timeout problem for a lot of folks!

**Comment by larsoner** [comment]
@mjg for your error I would open a bug with the VTK folks (of course after some searching), it's possible they will recognize something from the traceback:

https://gitlab.kitware.com/vtk/vtk/-/issues

Alternatively you could engage in a time consuming `git bisect` + wheel build + test to figure out the VTK commit that caused the problem, e.g. what pyvista folks did here:

https://github.com/pyvista/pyvista/pull/1394#issuecomment-865308257

**Review by prabhuramachandran — APPROVED** [review]
Thank @larsoner! This looks good to me, I will refactor the print debug statements in a separate PR.

**Comment by larsoner** [comment]
And this exists:
```
(Pdb) p tvtk.Version().get_vtk_version()
'9.1.0'
```
So I'm not sure where/why the traits-based wrapper is failing :(

**Comment by prabhuramachandran** [comment]
@larsoner -- I will merge this for now to make life easier for everyone, I will look at the additional failures for 9.1.0 tomorrow and will not be able to look at this today. Looking at the errors, it looks like 9.1.0 changed the docstring signature which is going to break many things for sure.  Thanks for bringing this up.  My first goal is to get 9.0.3 all set and the tests going there and then I will look at 9.1.0-rc2.  Am hoping that by next weekend things will be in better shape.

File: tvtk/vtk_parser.py [hunk 1]
```diff
@@ -640,7 +643,10 @@ def _find_get_set_methods(self, klass, methods):
 
         # Find the default and range of the values.
         if gsm:
+            # Useful for debugging on failures:
```
Comment (mdickinson) [inline tvtk/vtk_parser.py:646]: For future evolution of this code, how about using logging in place of the commented-out print statements?

Comment (prabhuramachandran) [reply tvtk/vtk_parser.py:646]: This code is primarily used at build time, so would logging be appropriate?  I was thinking of using an environment variable like "MAYAVI_DEBUG_BUILD" or so to turn on these print statements.  What do you recommend?

Comment (mdickinson) [reply tvtk/vtk_parser.py:646]: Not sure, but just from a code cleanliness point of view I'd really like to get away from having commented-out code in the codebase.

A "verbose" or "debug" flag to turn on the print statements doesn't seem unreasonable; that's probably better than using an environment variable at this point - we probably want the environment variable handling in the topmost layer.

**Comment by mdickinson** [comment]
@prabhuramachandran The inability to merge was purely related to the PR, not to you. I wasn't being allowed to merge either.  :-) Everyone in the "enthought" team in the "enthought" org currently has write privileges on this repository, which should allow merging PRs.

The merge was blocked because the expected GitHub Actions test runs hadn't passed, and the branch protections that we recently put in place require those test runs to pass. And the test runs didn't pass because they weren't running, which in turn was because the connected GitHub Actions workflow was present in the master branch but not in this branch. That's why the merge from master into this branch fixed things. We're likely to see the same effect with other open PRs that haven't had a merge from master recently. The fix should be the same for those PRs: merge master into the branch.

**Comment by mdickinson** [comment]
@prabhuramachandran Re: admin privileges, @rahulporuri should be able to push that through on our side.

**Pull request merged**